src/textual/_spatial_map.py

```diff
@@ -0,0 +1,94 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from itertools import product
+from typing import Generic, Iterable, TypeVar
+
+from .geometry import Region
+
+ValueType = TypeVar("ValueType")
+
+
+class SpatialMap(Generic[ValueType]):
+    """A spatial map allows for data to be associated with a rectangular regions
+    in Euclidean space, and efficiently queried.
```

Comment:
This could be improved because this docstring doesn't do justice to what you explained directly to me. 

My suggestion would be to consider writing a first, short sentence that gives a very high-level description of what this class is/represents, e.g., `"Provide fast spatial querying for data arranged in a Euclidean space."`.

Then, elaborate on what you mean by "allows for data to be associated with a rectangular region in Euclidean space". What type of association are you talking about?

On top of that, this is a "private" module, which means it is highly unlikely that we will write further documentation for this other than the one we leave in the docstrings, so it is a good idea to make it as comprehensive as possible.

Also, is this related/similar to [quadtrees](https://en.wikipedia.org/wiki/Quadtree)?